Device roles to capabilities (#197)
margo/specification: system-design/specification/margo-devices/device-requirements.md

# Device Requirements

> All Margo conformant devices MUST meet a baseline set of requirements to be considered Margo compliant. These requirements ensure devices can effectively manage Margo compliant workload and/or host those workloads. A device MUST support onboarding and management from only a single workload fleet manager. A device reports what it can provide to the ecosystem through its [device capabilities](../margo-management-interface/device-capabilities.md).

Reviewer: ```suggestion
All Margo conformant devices MUST meet a baseline set of requirements to be considered Margo compliant. These requirements ensure devices can effectively manage Margo-compliant workload and/or host those workloads. A device MUST support onboarding and management from only a single workload fleet manager. A device reports what it can provide to the ecosystem through its [device capabilities](../margo-management-interface/device-capabilities.md).
```

> All devices MUST provide the following components:

Reviewer: Consider qualifying the device as Margo-conformant, and then you can remove the note about devices behind a see-through gateway because those are not Margo-conformant.

```suggestion
All margo-conformant devices MUST provide the following components:
```

- A workload fleet management client

> Note: This is not required for a device behind a see-thru gateway.

All workload hosting devices MUST provide the following components:
    
- at least one supported deployment type (`supportedDeploymentTypes`)

> - at least one supported workload runtime (`supportedRuntimes`)

Reviewer: Were the `supportedDeploymentTypes` and `supportedRuntimes` intended to be links? Having these terms in the parentheses without any context feels odd. I'm not sure they are needed, but if you want to keep them, then maybe add something like:

"See the `supportedDeploymentTypes` in [device capabilities](../page.md) for the supported deployment types"

Author: Should be addressed in latest commit.

Reviewer: The added note is incomplete because it only addresses the deployment types.

>Note: See the supportedDeploymentTypes in [device capabilities](https://github.com/margo/specification/blob/ajc/roles-to-capabilities/system-design/specification/margo-management-interface/device-capabilities.md) for the supported deployment types.

I'd suggest. I don't think we need to call out the object names specifically.

>Note: See [device capabilities](https://github.com/margo/specification/blob/ajc/roles-to-capabilities/system-design/specification/margo-management-interface/device-capabilities.md) for the supported deployment types and runtimes.

Author: addressed in latest commit. I opted to remove the parameters from the bullet point. One less place ot update if we change, along with the link provided for readers to follow up on the supported deployment types/runtime.

- An OTEL collector

> ## Kubernetes (Helm) enabled device requirements

Reviewer: Is there a reason we want to call out "Kubernetes" here instead of just using "Helm"? If there is a reason you want Kubernetes here, it suggests we need additional content somewhere.

Author: Fixed in latest commit

Devices supporting the helm deployment type, MUST provide the following additional components:

- Capabilities for deploying Helm charts

>     - described via the `supportedDeploymentTypes`:`helm`

Reviewer: Same suggestion as above with the `supportedDeploymentTypes` and `supportedRuntimes`

Author: I've removed that sub bullet. Link is provided to the supported deployment types within the generic device requirements at the top.

# Device Requirements

> All Margo conformant devices MUST meet a baseline set of requirements to be considered Margo compliant. These requirements ensure devices can effectively manage Margo-compliant workload and/or host those workloads. A device MUST support onboarding and management from only a single workload fleet manager. A device reports what it can provide to the ecosystem through its [device capabilities](../margo-management-interface/device-capabilities.md).

Reviewer 2: 1. The first sentence reads circular :)
2. I'd align on a single wording for "Margo-conformant/-compliant" (I suggest Margo-conformant)

```suggestion
A devices MUST meet a baseline set of requirements to be considered Margo-conformant. These requirements ensure devices can effectively manage Margo-conformant workload and/or host those workloads. A device MUST support onboarding and management from only a single workload fleet manager. A device reports what it can provide to the ecosystem through its [device capabilities](../margo-management-interface/device-capabilities.md).
```

> All margo-conformant devices MUST provide the following components:

Reviewer 2: ```suggestion
All Margo-conformant devices MUST provide the following components:
```

- A workload fleet management client

> Note: This is not required for a device behind a see-thru gateway.

> All workload hosting devices MUST provide the following components:

Reviewer 2: ```suggestion
All workload hosting devices MUST provide the following capabilities:
```

    
- at least one supported deployment type

> - at least one supported workload runtime

Reviewer 2: ```suggestion
- At least one supported deployment type
- At least one supported workload runtime
```

- An OTEL collector

> Note: See [device capabilities](../margo-management-interface/device-capabilities.md) for the supported deployment types and runtimes. 

> ## Helm enabled device requirements

Reviewer 2: ```suggestion
## Helm-enabled Device Requirements
```

> Devices supporting the helm deployment type, MUST provide the following additional components:

Reviewer 2: ```suggestion
Devices supporting the Helm deployment type, MUST provide the following additional components:
```

If a device vendor chooses a deployment approach that does not interact with the Kubernetes API, the list of Kuberentes APIs available in the cluster MUST be provided out-of-band. This is supported by a variety of tooling such as the Helm GO SDK, `helm template` command, Kustomization, and ArgoCD.

## Compose enabled device requirements

Devices supporting the compose deployment type, MUST provide the following additional components:

- Capabilities for deploying Compose applications

>     - described via the `supportedDeploymentTypes`:`compose`

Reviewer: Same comment as above

Author: I've removed that sub bullet. Link is provided to the supported deployment types within the generic device requirements at the top.

If a device vendor chooses a deployment approach that does not interact with the Kubernetes API, the list of Kuberentes APIs available in the cluster MUST be provided out-of-band. This is supported by a variety of tooling such as the Helm GO SDK, `helm template` command, Kustomization, and ArgoCD.

> ## Compose enabled device requirements

Reviewer 2: ```suggestion
## Compose-enabled Device Requirements
```

> Devices supporting the compose deployment type, MUST provide the following additional components:

Reviewer 2: ```suggestion
Devices supporting the Compose deployment type, MUST provide the following additional components:
```

- Capabilities for deploying Compose applications

> - Compose orchestration platform i.e., Docker or Podman

Reviewer 2: Should we say "Compose-conformant"?

An additional thought unrelated to this PR: should we pin a Compose spec version? And/or allow client to report the version they conform to? Same for Kubernetes...

Author: Are you stating that for the "capabilities for deploying compose applications"? 

Or

was this regarding the orchestration platform note?

Reviewer 2: Mh, this actually was related to the orchestration platform note. But thinking more about it, "Compose-conformant orchestration platform" makes it even worse :) What we are actually trying to say here is probably just "OCI container runtime".

```suggestion
- An OCI container runtime, for example Docker or Podman
```
